### ISSUE TITLE
add CHANGELOG.md file

### ISSUE THREAD

**Issue by bdarcus** [citation-style-language/schema#220]
@rmzelle has mentioned this in the past, and it's time we figure it out.

Obvious is to add a [CHANGELOG.md file](https://github.com/citation-style-language/schema/blob/master/CHANGELOG.md), which I have just done.

Question is how to populate it, and maintain it.

One option going forward is to automate based on the commit history.

Please see these two links:

- https://keepachangelog.com/en/1.0.0/ (general)
- https://www.freecodecamp.org/news/a-beginners-guide-to-git-what-is-a-changelog-and-how-to-generate-it/ (discussion of tool options)

**Comment by denismaier** [comment]
Yeah, a changelog is necessary. I'm not sure doing this automatically is the best way in that case.

**Comment by rmzelle** [comment]
See https://docs.citationstyles.org/en/1.0.1/release-notes.html and https://docs.citationstyles.org/en/1.0/release-notes.html for the existing changelogs.

**Comment by denismaier** [comment]
Question @bdarcus: Is this about a changelog for the released version (like the linked release notes)? Or about a work in progress log?

**Comment by bdarcus** [comment]
> Question @bdarcus: Is this about a changelog for the released version (like the linked release notes)? Or about a work in progress log?

Both, per recommendation on the "keep" site.

I just didn't add the earlier stuff before we settle details.

**Comment by bdarcus** [comment]
> See https://docs.citationstyles.org/en/1.0.1/release-notes.html and https://docs.citationstyles.org/en/1.0/release-notes.html for the existing changelogs.

Oh wow; those are quite detailed!

I was hoping to just use something like this, at least as a started point, with links to issue and PRs.

``` console
❯ git log --pretty="- %s"
- Add content to CHANGELOG
- Create CHANGELOG.md
- Use variables.standard pattern, format
- Add trang script for rnc validation and formatting
- Simplify and add structure to csl.rnc formatting (#210)
- Add .rng to .gitgnore
- Add terms for "no-place" and "no-name" (#206)
- Add "document" type (#207)
- Add new item types; classic, hearing, etc (#194)
- Add additional locator variables (#201)
```

... which of course would convert to this in rendered markdown:

- Add content to CHANGELOG
- Create CHANGELOG.md
- Use variables.standard pattern, format
- Add trang script for rnc validation and formatting
- Simplify and add structure to csl.rnc formatting (#210)
- Add .rng to .gitgnore
- Add terms for "no-place" and "no-name" (#206)
- Add "document" type (#207)
- Add new item types; classic, hearing, etc (#194)
- Add additional locator variables (#201)

The better and more consistent our commit and commit message approach, the easier this is to manually maintain.

To go my question, then, on the format, one question is whether is follow a flatter, traditional, changelog format, or do something more structured.

The former is easier to manage.

**Comment by bdarcus** [comment]
Here's [an example](https://gist.github.com/bdarcus/e63cad6748de8af764744272241b6dbb) using `auto-changelog`, without any configuration (though a lot is available).

**Comment by rmzelle** [comment]
> Oh wow; those are quite detailed!

I think the target audience of a CSL schema/specification changelog should include CSL style authors, CSL processor implementors, and downstream implementations like Zotero. So I would try to spend some time to create something that's accessible.

(and IIRC, I wrote the original changelogs once each release was feature-frozen, by trawling through all the commits to make sure I hadn't missed anything, so a tool like "auto-changelog" might certainly come in handy and help link the changes to their PRs)

**Comment by bdarcus** [comment]
In that case, maybe we can maintain a clean, developer-oriented, changelog file here, created with auto-changelog, but manually-edited where we need.

You (and whoever else might have time and interest) can then build wider-facing release note when time permits, based on that repo changelog (which as you note will include important links and such).

I've added the config file I used to the repo, and updated the file to use its output. When we do the releases, we can fix the content.

We can also leave this issue open as we figure out the details (on customization, and on commit message conventions going forward, etc).

**Comment by bdarcus** [comment]
We probably want a [custom template](https://github.com/CookPete/auto-changelog#custom-templates) to organize the output for our needs.

**Comment by bdarcus** [comment]
draft; trying to think this through ATM.

<details>
Proposal, for commit message conventions that will support auto changelog maintanance.

### Summary Format

The basic template would be `[feature]: [Keyword] [short detail] (GH PR number]`.

An empty keyword assumes the CSL rnc schemas.

Examples: 

- `input: Add 'foo' variable`
- `Add 'bar` variable`
- `Deprecate A`
- `CI: Add validation for X`

### Keywords 

- add
- deprecate
</details>

**Comment by denismaier** [comment]
> * **deprecate** (patterns or strings dropped from schemas, with processing implications)

We'll need to settle on the meaning of "deprecate"... Is it: "Removed and invalid from now on." Or: "Don't use this no longer as we'll remove this in an upcoming release."
(I'd say it is the latter, and I think that was also @bwiernik's take on this. Right?)

**Comment by bdarcus** [comment]
Good point.

As I've been using it in commit messages so far, it is the former.

But I'm not wedded to that. I agree, we need to decide.

**Comment by bwiernik** [comment]
Yes, it’s the latter. Deprecate means “still here for compatibility, but don’t use”. If it is being removed from the schema/spec, just remove it.

**Comment by bdarcus** [comment]
OK, I updated the draft, but keeping as draft, since I'm still thinking about how best to do this.

**Comment by bdarcus** [comment]
So also this github action, which would allow us to integrate auto-labeling of issues and PRs based on keywords as well.

https://github.com/marketplace/actions/issue-auto-labeling-and-assigning

**Comment by bwiernik** [comment]
The change log is for people who aren't closely following the development. It is fine for it to be somewhat redundant with the commit log--we shouldn't expect someone writing or updating a citation processor to read through commit logs.

**Comment by bdarcus** [comment]
Exactly. What I'm saying is if adopt the right convention for commit log messages, then the auto-updated changelog will actually be useful for those folks, and the auto-label assignment should also just work.
